### ISSUE TITLE
Needs different flags on macOS

### ISSUE THREAD

**Issue by xvxx** [Lartu/ldpl-ncurses#2]
[Lute](https://github.com/Lartu/lute) was failing to compile for me so I tracked it down to the fact that macOS doesn't want `flag "-lncursesw"` but instead needs:

````coffeescript
flag "-D_XOPEN_SOURCE_EXTENDED"
flag "-lncurses"
````

Is there a way to set per-os flags using the `flag` statement? Should there be `linux flag` and `mac flag` and so on...?

**Comment by Lartu** [comment]
My deepest fears have come true, having to do things in different ways in different OSs! There is no way at the moment to choose flags depending on the OS. And, if we added something like `mac flag`, we would have to add special flag statements for each os 😱

Maybe we could add something like:

```
flag set 1:
  flag "-lLinux"
flag set 2:
   flag "-lmacos"
   flag "-open_source_something"
default flag set:
   #...
```

So that you could compile using an `-option` flag that would choose one or another.

For example if I wanted to compile for macOS I'd do: 
`ldpl mysource.ldpl -flagset=2`

Maybe instead of using a default, flag set 1 would be used by default.

At the moment the only way we have is to mention on the readme what changes one should perform in order to be able to compile the project on each os.

**Comment by xvxx** [comment]
> My deepest fears have come true, having to do things in different ways in different OSs!

😱😱😱  

> And, if we added something like mac flag, we would have to add special flag statements for each os

That's true, but we could always start small. Go keeps a list of OSes:

https://github.com/golang/go/blob/6ba0be163/src/go/build/syslist.go#L10

However I understand not wanting to maintain a list of OSes, plus maybe there are other cases of optional flags that aren't OS specific. So I like the idea of setting flags, maybe they can be named instead of numbered?

````
flag set default "-LLinux"
flag set mac "-lmacos"
flag set mac "-open_source_something"
flag set win "-lWIN32"
flag "-prod"
````

Then: 

    ldpl mysource.ldpl -flag=mac

This would add `-lmacos -open_source_something -prod` to the c++ compile line, since `-prod` wasn't given a flag name it would always be included.

**Comment by xvxx** [comment]
Just came across this today - looks like the V language has per-OS preprocessor flags:

https://github.com/nsauzede/vsdl2/blob/2610c225f74a89dbc662f83a13e6aefab4bf2c2f/vsdl2.v#L7-L17

Would something like that work for you if it had a simple implementation, or would you rather go with name/number flags?

**Comment by Lartu** [comment]
I love that, but how would you take into consideration other systems like BSD or Plan9? Or other linux distros? What determines if an operating system is a "Linux"? I'm afraid the environmental variables and system files that state what operating system you are in might change and break everything for us.

I mean, being able to use `FLAG (LINUX) "-fSDL2"` would be wonderful, but I do fear being able to detect the OS in a trustworthy way might not be a simple task to complete.

**Comment by xvxx** [comment]
Fixed by #4
